### PR TITLE
prov/tcpx: Fix wrong error handling in fi_passive_ep()

### PR DESCRIPTION
**Pull request by sydidelot**
Lifbabric doesn't report the correct error code to the
caller if fi_passive_ep() fails.

For example, in the case where the port is already bound
to another process, fi_passive_ep() would return -1 to
the application. It's because the value returned by bind()
is forwarded to the caller instead of the errno value.
(on error, bind() returns -1, and errno is set appropriately).

The patch also fixes an issue where fi_dupinfo() may fail
but fi_passive_ep() would return FI_SUCCESS.

Signed-off-by: Sylvain Didelot <sdidelot@ddn.com>

### CODE REVIEW

File: prov/tcp/src/tcpx_ep.c [hunk 1]
```diff
@@ -199,13 +199,13 @@ static int tcpx_bind_to_port_range(SOCKET sock, void* src_addr, size_t addrlen)
...
 		ret = bind(sock, src_addr, (socklen_t) addrlen);
 		if (ret) {
-			if (errno == EADDRINUSE)
+			if (ofi_sockerr() == EADDRINUSE)
 				continue;
 
 			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
 				"failed to bind listener: %s\n",
 				strerror(ofi_sockerr()));
-			return -errno;
```
Comment (aakefbs) [inline prov/tcp/src/tcpx_ep.c:208]: I would store errno, into another var as FI_WARN() has the potential risk to change errno. Hmm, the other returns above have the same issue.

File: prov/tcp/src/tcpx_ep.c [hunk 2]
```diff
@@ -779,7 +781,10 @@ int tcpx_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
 
 	_pep->info = fi_dupinfo(info);
 	if (!_pep->info)
+	{
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c:784]: nit: Can you place { on same line as if.

File: prov/tcp/src/tcpx_ep.c (hunk 1 above)
Comment (shefty) [reply prov/tcp/src/tcpx_ep.c:208]: Have you see errno change with FI_WARN?  I didn't think the fi_log path calls anything that sets errno.

File: prov/tcp/src/tcpx_ep.c [hunk 3]
```diff
@@ -239,10 +239,12 @@ static int tcpx_pep_sock_create(struct tcpx_pep *pep)
...
 		goto err;
 	}
-	if (ofi_addr_get_port(pep->info->src_addr) != 0 || port_range.high == 0)
+	if (ofi_addr_get_port(pep->info->src_addr) != 0 || port_range.high == 0) {
 		ret = bind(pep->sock, pep->info->src_addr,
 			  (socklen_t) pep->info->src_addrlen);
-	else
+		if (ret)
+			ret = -ofi_sockerr();
+	} else
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c:247]: nit: please add braces around else clause, when if clause also needs it.

File: prov/tcp/src/tcpx_ep.c (hunk 1 above)
Comment (aakefbs) [reply prov/tcp/src/tcpx_ep.c:208]: I didn't get it with FI_WARN() yet, but I got overwritten errnos in my long distant past from printf functions. It is not mentioned in the linux man page, but at least opengroup states

<quote>
If an output error was encountered, these functions shall return a negative value [CX] [Option Start]  and set errno to indicate the error. 
</quote>

https://pubs.opengroup.org/onlinepubs/9699919799/functions/fprintf.html

Comment (shefty) [reply prov/tcp/src/tcpx_ep.c:208]: Hmm... I think the safest option here would be to modify the low-level fi_log() call to save and restore errno.  My guess is there are dozens of places in the code where errno is returned after calling one of the FI_LOG functions.  This shouldn't impact performance.

Comment (aakefbs) [reply prov/tcp/src/tcpx_ep.c:208]: Ok fine, going to create a patch for that.

Comment (sydidelot) [reply prov/tcp/src/tcpx_ep.c:208]: Let me know if you want me to do the patch

Comment (sydidelot) [reply prov/tcp/src/tcpx_ep.c:208]: Let me know me if you want me to do the patch. Thanks!

**Pull request merged**